db/crud.go

```diff
@@ -727,6 +710,39 @@ func (db *Database) PutExistingRev(docid string, body Body, docHistory []string)
 	return err
 }
 
+// Allow-conflicts=false handling
+func (db *Database) IsIllegalConflict(doc *document, parentRevID string, deleted bool) bool {
+	if db.AllowConflicts() {
+		return false
+	}
+
+	// Conflict-free mode: If doc exists, its current rev must be the new rev's parent, unless it's a tombstone.
+
+	// If the parent is the current rev, it's not a conflict.
+	if parentRevID == doc.CurrentRev || doc.CurrentRev == "" {
+		return false
+	}
+
+	// If the parent isn't the current rev, reject as a conflict unless this is tombstoning an existing non-winning leaf
+	if !deleted {
+		base.LogTo("CRUD+", "Conflict - non-tombstone updates to non-winning revisions aren't valid when allow_conflicts=false")
+		return true
+	}
+
+	// If it's a tombstone, it's only allowed if it's tombstoning an existing non-tombstoned leaf
+	illegalConflict := true
```

Comment:
Could you `return false` inside the loop, and below log and `return true`?

It avoids the cognitive overhead of tracking the state of `illegalConflict` when reading the code, and matches the early-return style above.

```go
	// If it's a tombstone, it's only allowed if it's tombstoning an existing non-tombstoned leaf
	for _, leafRevId := range doc.History.GetLeaves() {
		if leafRevId == parentRevID && doc.History[leafRevId].Deleted == false {
			return false
		}
	}

	base.LogTo("CRUD+", "Conflict - tombstone updates to non-leaf or already tombstoned revisions aren't valid when allow_conflicts=false")
	return true 
```

Reply:
I'd implemented it that way originally, and changed it when I thought it might not be clear what's going on.  Sounds like the current approach isn't any more clear, so I'll switch it back - thanks.